service/src/main/kotlin/no/nav/su/se/bakover/service/behandling/BehandlingServiceImpl.kt

```diff
@@ -0,0 +1,171 @@
+import java.time.LocalDate
+import java.util.UUID
+
+internal class BehandlingServiceImpl(
+    private val behandlingRepo: BehandlingRepo,
+    private val hendelsesloggRepo: HendelsesloggRepo,
+    private val beregningRepo: BeregningRepo,
+    private val oppdragRepo: OppdragRepo,
+    private val simuleringClient: SimuleringClient,
+    private val utbetalingService: UtbetalingService, // TODO use services or repos? probably services
+    private val oppgaveClient: OppgaveClient,
+    private val utbetalingPublisher: UtbetalingPublisher,
+    private val søknadService: SøknadService,
+    private val sakService: SakService
+) : BehandlingService {
+    override fun hentBehandling(behandlingId: UUID): Either<FantIkkeBehandling, Behandling> {
+        return behandlingRepo.hentBehandling(behandlingId)?.right() ?: FantIkkeBehandling.left()
+    }
+
+    override fun underkjenn(
+        begrunnelse: String,
+        attestant: Attestant,
+        behandling: Behandling
+    ): Either<Behandling.KunneIkkeUnderkjenne, Behandling> {
+        return behandling.underkjenn(begrunnelse, attestant)
+            .map {
+                hendelsesloggRepo.oppdaterHendelseslogg(it.hendelseslogg)
+                behandlingRepo.hentBehandling(it.id)!!
+            }
+    }
+
+    override fun oppdaterBehandlingsinformasjon(
+        behandlingId: UUID,
+        behandlingsinformasjon: Behandlingsinformasjon
+    ): Behandling {
+        val beforeUpdate = behandlingRepo.hentBehandling(behandlingId)!!
+        beregningRepo.slettBeregningForBehandling(behandlingId)
+        val updated = behandlingRepo.oppdaterBehandlingsinformasjon(
+            behandlingId,
+            beforeUpdate.behandlingsinformasjon().patch(behandlingsinformasjon)
+        )
+        // TODO fix weirdness for internal state
+        val status = updated.oppdaterBehandlingsinformasjon(behandlingsinformasjon).status()
+        behandlingRepo.oppdaterBehandlingStatus(behandlingId, status)
+        return behandlingRepo.hentBehandling(behandlingId)!!
+    }
+
+    override fun opprettBeregning(
+        behandlingId: UUID,
+        fraOgMed: LocalDate,
+        tilOgMed: LocalDate,
+        fradrag: List<Fradrag>
+    ): Behandling {
+        beregningRepo.slettBeregningForBehandling(behandlingId)
+        val beregnet = behandlingRepo.hentBehandling(behandlingId)!!.opprettBeregning(fraOgMed, tilOgMed, fradrag)
+        beregningRepo.opprettBeregningForBehandling(behandlingId, beregnet.beregning()!!)
+        behandlingRepo.oppdaterBehandlingStatus(behandlingId, beregnet.status())
+        return behandlingRepo.hentBehandling(behandlingId)!!
+    }
+
+    override fun simuler(behandlingId: UUID): Either<SimuleringFeilet, Behandling> {
+        val behandling = behandlingRepo.hentBehandling(behandlingId)!!
+        val sak = sakService.hentSak(behandling.sakId).fold(
+            { throw RuntimeException("Kunne ikke finne sak") },
```

Comment:
Kunne vel gjort om SimuleringFeilet til en sealed class og kalt den KunneIkkeSimulere, hvor du har 2 objects, en for FantIkkeSak som gir 404 og en som heter SimuleringFeilet som gir 500

Reply:
Nvm, her burde man kanskje hatt hentSakForBehandlingId, og en FantIkkeBehandling i KunneIkkeSimulere